docs/decisions/0024-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,366 @@
+
+With the bottom-up approach, we need to add model information to `IResultBase`:
+
+```csharp
+/// <summary>
+/// Interface for model results
+/// </summary>
+public interface IResultBase
+{
+  /// <summary>
+  /// Model name or Id
+  /// </summary>
+  string ModelId { get; }
+
+  /// <summary>
+  /// Gets the model result data.
+  /// </summary>
+  ModelResult ModelResult { get; }
+}
+```
+
+We also need to propagate the model Id up via KernelResult:
+
+```csharp
+// In AIFunctionResultExtensions.cs
+...
+/// <summary>
+/// Function model id key for <see cref="ModelResult"/> records.
+/// </summary>
+public const string ModelIdKey = "ModelId";
+...
+/// <summary>
+/// Returns the model Id from <see cref="FunctionResult"/> metadata.
+/// </summary>
+/// <param name="result">Instance of <see cref="FunctionResult"/> class.</param>
+public static string? GetModelId(this FunctionResult result)
+{
+  if (result.TryGetMetadataValue(ModelIdKey, out string? modelId))
+  {
+    return modelId;
+  }
+
+  return null;
+}
+
+
+// In KernelFunctionFromPrompt.cs
+...
+result.Metadata.Add(AIFunctionResultExtensions.ModelIdKey, completionResults[0].ModelId);
+...
+return result;
+```
+
+A kernel function will then retrieve the information by doing the following:
+
+```csharp
+// Note that not all services support usage details.
+if (result.GetModelId() is null)
+{
+  logger.LogInformation("Model id not found in function result.");
+  return;
+}
+var modelId = result.GetModelId();
+
+if (result.GetModelResults() is null)
+{
+  logger.LogInformation("Model results not found in function result.");
+  return;
+}
+var modelResult = result.GetModelResults().FirstOrDefault();
+var modelResultJson = modelResult?.GetJsonResult();
+
+var promptTokens = 0;
+var completionTokens = 0;
+try
+{
+  var tokenUsage = modelResultJson.GetValueOrDefault().GetProperty("Usage");
+  promptTokens = tokenUsage.GetProperty("PromptTokens").GetInt32();
+  completionTokens = tokenUsage.GetProperty("CompletionTokens").GetInt32();
+}
+catch (Exception ex) when (ex is KeyNotFoundException)
+{
+  logger.LogInformation("Usage details not found in model result.");
+  return;
+}
+catch (Exception ex)
+{
+  logger.LogError(ex, "Error while parsing usage details from model result.");
+  throw;
+}
+```
+
+## To optionally allow the kernel to send sensitive information
+
+In some environments, it is not inappropriate to store sensitive information, such as PII, in any services. Nonetheless, those intermediate results (generated using non-customer data) can be extremely useful for debugging purposes. Thus, we need a switch that can turn on and off telemetries that contain sensitive information.
+
+```csharp
+// In a new InstrumentationOptions.cs
+public sealed class InstrumentationOptions
+{
+  public bool IncludeSensitiveInformation { set; get; } = false;
```

Comment:
I wonder how this property will be used. Do we plan to wrap code for writing logs/traces within if statements throughout the code? Or will it be done only in one location?

Reply:
We will wrap code for writing logs/traces within if statements. Please see an example in the implementation detail.